### DISCUSSION TITLE
Add one click code modification function using Spotless plugin

### DISCUSSION THREAD

**Discussion by Yuukadesu** [apache/streampipes#3170] [Dev]
Each contributor's code style is different. In order to ensure that the code style of streampipes meets the prescribed standards, I added the Spotless plugin, which can be modified to the specified code style with just one click.

**Comment by luoluoyuyu** [comment]
Hi @Yuukadesu 
I think this is a great idea, thanks for the discussion, would you like to contribute code to this idea?

**Comment by bossenti** [comment]
Do we want to abandon Checkstyle then?

**Reply by Yuukadesu** [reply]
Hi @bossenti 
It does not abandon the existing code style checking plug-in format, but only uses spotless to help contributors automatically fix the submitted code. https://github.com/diffplug/spotless

**Reply by bossenti** [reply]
Hm, can we run spotless based on the Checkstyle rules? I want to avoid to specify our rule set twice

**Reply by dominikriemer** [reply]
I would prefer spotless over checkstyle since it's more modern and more similar to our toolchain for the UI, where we also have automatic code style fixes. Also, I'm often running into issues with the manual approach used by checkstyle.

**Reply by tenthe** [reply]
I am completely open when it comes to tools. It's just important to me that it's easy for developers to set up the tooling and that we document how to do it. And we shouldn't have two tools that do the same thing

**Reply by bossenti** [reply]
I'm totally open for spotless too
Just want to make sure that we have a clear decision and don't inflate our dev setup :)

**Reply by luoluoyuyu** [reply]
I don't think we need to abandon the current Checkstyle plugin, we can just use Spotless to automatically fix the code

**Reply by Yuukadesu** [reply]
Currently, spotless does not integrate checkstyle, so I need to imitate the existing checkstyle.xml to rewrite an identical rule for spotless to use.
In view of the above, I have two ideas. One is to abandon checkstyle and use spotless, completely copying the code style rules defined by checkstyle.xml. However, this may cause significant changes; Another approach is to use both, converting checkstyle rules into spotless compatible rules and placing them in another XML file.Checkstyle is responsible for checking the code style, while Spotless is responsible for one click modification of the code style.

**Comment by bossenti** [comment]
>  I need to imitate the existing checkstyle.xml to rewrite an identical rule for spotless to use.

That's exactly what I would like to avoid. You then always need to keep them in sync and adapt them as soon as any of the tool changes it's behavior.

>  One is to abandon checkstyle and use spotless, completely copying the code style rules defined by checkstyle.xml.

This is completely fine for me 🙂 If we find some odd rules in the checkstyle config we can even get rid of them and adopt better patterns

**Reply by luoluoyuyu** [reply]
+1 I am looking forward to the introduction of Spotless, this plugin will make it easy for us to fix the code format

**Reply by Yuukadesu** [reply]
Due to the significant differences between the formats supported by checkstyle and spotless, such as some regular expressions. So I plan to use both, Spotless is responsible for automatically correcting the formatting of most of the code, while some with regular expressions will continue to use checkstyle constraints.

**Comment by Yuukadesu** [comment]
Hi! @dominikriemer @bossenti @luoluoyuyu @tenthe
I only added the Spotless plugin, but there were many code changes made.l hope someone familiar with the project can help me complete the code style modification together.https://github.com/apache/streampipes/pull/3290
